### PR TITLE
Server example with API Rest

### PR DESCRIPTION
**Pull request by FSSRepo**
Hi. These days, I have been working on adding an API to llama.cpp using [cpp-httplib](https://github.com/yhirose/cpp-httplib). I think it could help some people to implement Llama in their projects easily. I know that there are already several alternatives such as bindings to Node JS and Python, but with this example, I intend to implement it natively in C++.

For now, it can only be compiled with CMake on Windows, Linux, and MacOS. All usage and API information can be found in the `README.md` file inside the `examples/server` directory. It doesn't require an external dependency.

Edit:

Current available features:

- [x] Completion (wait to end, in loop)
- [x] Custom prompt (generation and interactive behavior)
- [x] Tokenize
- [x] Embeddings

Any suggestions or contributions to this PR are welcome.

### CODE REVIEW

**Comment by CRD716** [comment]
I feel like this goes pretty directly against the no dependencies rule.

**Comment by FSSRepo** [comment]
@CRD716 why?

**Comment by CRD716** [comment]
Because this adds extra headers and an external library in the form of cpp-httplib.
> Avoid adding third-party dependencies, extra files, extra headers, etc.
- Readme

**Comment by FSSRepo** [comment]
So, would it be better if this is a separate project that will never be part of the master repository? Is there no other way? Or should I just close this PR?

**Comment by CRD716** [comment]
I personally think this would be better as a separate project, but it's really up to ggerganov on if this is acceptable within the examples or not.

**Comment by prusnak** [comment]
@CRD716 this has been already discussed in https://github.com/ggerganov/llama.cpp/pull/1025 and we wan't to have a server example in this repository

**Comment by ggerganov** [comment]
Yes, what @prusnak says

In general, for this kind of examples that depend on something extra, make sure to:

- Not add them to the `Makefile`. Use only `CMake` and put the example behind a CMake option that is disabled by default
- Put all 3rd party dependencies (i.e. json lib, etc.) inside the new folder in the `examples`. They will be used only by that example
- If possible, provide a separate CI job for that example so that it gets long-term support when breaking changes occur

Keep in mind that we might decide to remove such examples at any point if the maintenance effort becomes too big and there is nobody to do it

An alternative approach is to make a fork and add a README `example` in `llama.cpp` that describes the purpose of the example and links to the fork. This way, the maintenance efforts will be up to the owner of the fork

---

Note that the rules in the README refer to third-party dependencies to the core `llama` and `ggml` code.
The examples can link to small dependencies, but we still have to be thoughtful about what we link and choose the minimal option

**Comment by FSSRepo** [comment]
> If we really want JSON support, maybe we can use https://github.com/nlohmann/json which is a) maintained and b) single-file include only library?

I will change it

File: examples/server/server.h
Comment (SlyEcho) [inline examples/server/server.h:1]: Why is this file necessary and why does it include headers it doesn't use?

File: examples/server/server.cpp [hunk 1]
```diff
@@ -0,0 +1,794 @@
...
+}
+
+void server_print_usage(int /*argc*/, char ** argv, const gpt_params & params) {
+    fprintf(stderr, "usage: %s [options]\n", argv[0]);
+    fprintf(stderr, "\n");
+    fprintf(stderr, "options:\n");
+    fprintf(stderr, "  -h, --help            show this help message and exit\n");
+    fprintf(stderr, "  -s SEED, --seed SEED  RNG seed (default: -1, use random seed for < 0)\n");
+    fprintf(stderr, "  --memory_f32          use f32 instead of f16 for memory key+value\n");
+    fprintf(stderr, "  --keep                number of tokens to keep from the initial prompt (default: %d, -1 = all)\n", params.n_keep);
```
Comment (SlyEcho) [inline examples/server/server.cpp:513]: `n_keep` should be passed to the API, since it determines how any given input is evaluated.

File: examples/server/server.cpp [hunk 2]
```diff
@@ -0,0 +1,794 @@
...
+                    llama->params.prompt = ctx_msg["content"].get<std::string>() + "\n\n";
+                  }
+                  else if (role == "user")
+                  {
+                    llama->params.prompt += llama->user_tag + " " + ctx_msg["content"].get<std::string>() + "\n";
+                  }
+                  else if (role == "assistant")
+                  {
+                    llama->params.prompt += llama->assistant_tag + " " + ctx_msg["content"].get<std::string>() + "\n";
+                  }
```
Comment (SlyEcho) [inline examples/server/server.cpp:699]: This formatting seems too fixed, with the spaces and newlines. I don't think it can work for Alpaca at all.

What happens if the role is not `system` or `user` or `assistant`?

**Comment by SlyEcho** [comment]
What is this API based on? Does it follow some existing standard or service so that it could be used with other software?

Overall, there is a lot of code dealing with prompt generation, when there should just really be two: evaluate text, generate text (with stop keyword). It's not that hard to generate prompt from a past chat, for example, in Javascript.

There could also be a tokenization endpoint so that the client can estimate the need for truncation or summarization.

But these are my opinions.

File: examples/server/server.h
Comment (FSSRepo) [reply examples/server/server.h:1]: I'm thinking of removing it and declaring the functions directly as global functions in `server.cpp`, and creating a struct with all the required parameters for it to work instead of using `gpt_params`. That way, only `httplib.h`,` json.hpp`, and `llama.h` will be included in `server.cpp`.

File: examples/server/server.cpp (hunk 1 above)
Comment (FSSRepo) [reply examples/server/server.cpp:513]: You're right

File: examples/server/server.cpp (hunk 2 above)
Comment (FSSRepo) [reply examples/server/server.cpp:699]: Initially, this server implementation was more geared towards serving as a chatbot-like interaction. It was roughly based on the OpenAI [ChatGPT API](https://platform.openai.com/docs/guides/chat) although there were unfortunately many limitations such as the fact that prompt evaluation is a bit slow, so the initial context is constant from the beginning. It is also not possible to reset the context (without the need to reload the model) or at least I haven't found a way to do so.

**Comment by FSSRepo** [comment]
Initially, this server implementation was more geared towards serving as a chatbot-like interaction. It was roughly based on the OpenAI [ChatGPT API](https://platform.openai.com/docs/guides/chat) although there were unfortunately many limitations such as the fact that prompt evaluation is a bit slow, so the initial context is constant from the beginning. It is also not possible to reset the context (without the need to reload the model) or at least I haven't found a way to do so.

File: examples/server/server.cpp [hunk 3]
```diff
@@ -0,0 +1,794 @@
...
+
+int main(int argc, char ** argv) {
+
+  // own arguments required by this example
+  gpt_params default_params;
+
+  gpt_params params;
+  params.model = "ggml-model.bin";
+
+  std::string hostname = "0.0.0.0";
```
Comment (slaren) [inline examples/server/server.cpp:537]: It may be better to only listen on 127.0.0.1 by default, exposing the server to the network can be dangerous.

File: examples/server/server.cpp (hunk 2 above)
Comment (FSSRepo) [reply examples/server/server.cpp:699]: When `role` is not `user`, `assistant` or `system` the content is ignored.

Related Alpaca model, passing the following options in setting-context should work:

```javascript
const axios = require('axios');

async function Test() {
    let result = await axios.post("http://127.0.0.1:8080/setting-context", {
        context: [
            { role: "system", content: "Below is an instruction that describes a task. Write a response that appropriately completes the request." },
            { role: "user", content: "Write 3 random words" },
            { role: "assistant", content: "1. Sunshine\n2. Elephant\n3. Symphony" }
        ],
        tags: { user: "### Instruction:", assistant: "### Response:" },
        batch_size: 256,
        temperature: 0.2,
        top_k: 40,
        top_p: 0.9,
        n_predict: 2048,
        threads: 5
    });
    result = await axios.post("http://127.0.0.1:8080/set-message", {
        message: 'How to do a Hello word in C++, step by step'
    });
    if(result.data.can_inference) {
		result = await axios.get("http://127.0.0.1:8080/completion?stream=true", { responseType: 'stream' });
        result.data.on('data', (data) => {
            let dat = JSON.parse(data.toString());
            // token by token completion
            process.stdout.write(dat.content);
            if(dat.stop) {
                console.log("Completed");
            }
        });
    }
}

Test();
```

Output:
```bash
node test.js


1. Include the <iostream> header file
2. Use the "cout" object to output the text "Hello World!"
3. Add a new line character at the end of the output
4. Display the output on the screen

Example code:
'''c++
#include <iostream>

int main() {
    cout << "Hello World!" << endl;
    return 0;
}
'''
Completed
```

File: examples/server/server.cpp (hunk 3 above)
Comment (FSSRepo) [reply examples/server/server.cpp:537]: You're right. I will fix it.

File: examples/server/server.cpp (hunk 2 above)
Comment (SlyEcho) [reply examples/server/server.cpp:699]: I mean it will actually format it as `### Instruction: Write 3 random words\n### Response: 1. Sunshine\n2. Elephant\n3. Symphony` when [Alpaca is trained](https://github.com/tloen/alpaca-lora/blob/main/templates/alpaca.json) on `### Instruction:\nWrite 3 random words\n\n### Response:\n1. Sunshine\n2. Elephant\n3. Symphony`. It does work but that's because LLMs are flexible, but it's not optimal.

Don't get me wrong, I am trying to look out for you, if you decide to add this prompt templating into the API then you will be fixing and adding features to it for a long time when people come up with new models and ideas.

How about this?

```javascript
const context = [
    { user: "Write 3 random words", assistant: "1. Sunshine\n2. Elephant\n3. Symphony" }
]
const system = 'Below is an instruction that describes a task. Write a response that appropriately completes the request.'
const message = 'How to do a Hello word in C++, step by step'
const prompt = `${system}

${context.map(c => `### Instruction:
${c.user}

### Response:
${c.assistant}`)}

### Instruction:
${message}
`

// I think these two calls could be just merged into one endpoint
await axios.post("http://127.0.0.1:8080/setting-context", {
    content: prompt,
    batch_size: 256, // actually these could be command line arguments, since
    threads: 5,      // they are relevant to the server CPU usage
})
let result = await axios.get("http://127.0.0.1:8080/completion?stream=true", {
    responseType: 'stream', 
    temperature: 0.2,  // sampling parameters are used only for
    top_k: 40,         // generating new text
    top_p: 0.9,
    n_predict: 2048,
    
    // this is why it's nice to be able to get tokens
    keep: (await axios.post("http://127.0.0.1:8080/tokenize", { content: system })).length,
    // alternative is to pass keep as a string itself, but then you need to check if its tokens match what was given beforebegin
    
    stop: ['###', '\n'], // stop generating when any of these strings is generated,
                         // when streaming you have to make sure that '#' or '##' is not
                         // shown to the user because it could be a part of the stop keyword
                         // so there needs to be a little buffer.
})
let answer = ''
result.data.on('data', (data) => {
    let dat = JSON.parse(data.toString())
    // token by token completion
    process.stdout.write(dat.content)
    answer += dat.content
    if(dat.stop) {
        console.log("Completed")
        // save into chat log for the next question
        context.push({ user: message, assistant: answer })
    }
})
```

Actually there are libraries for this and other kinds of prompts in [langchain](https://js.langchain.com/docs/modules/prompts/prompt_templates/additional_functionality#few-shot-prompt-templates) for example.

Comment (FSSRepo) [reply examples/server/server.cpp:699]: I have a problem with `cpp-httplib`. It seems that your current implementation only allows streaming of data with GET requests. You attempted to create a single 'completion' endpoint of type POST to allow for defining options, and it works. However, it doesn't work when you try to implement streaming, and you're unsure of the reason.

Comment (FSSRepo) [reply examples/server/server.cpp:699]: I have a problem with cpp-httplib. And it's because it only allows streaming of data with GET requests. I tried to create a single POST endpoint called 'completion' to allow defining options, and it works. However, when it comes to streaming, it doesn't work for some reason.

Comment (FSSRepo) [reply examples/server/server.cpp:699]: The idea you propose seems interesting, but there are limitations such as the time it takes to evaluate the prompt, especially if the prompts are very long. It also involves restarting the prompt, and I'm not sure how that could be done.

Comment (FSSRepo) [reply examples/server/server.cpp:699]: Currently in this server, the `setting-context` needs to be called only once, `set-message` and `completion` need to be called whenever a response is desired, that allows for fast responses. The interaction context is being stored in "embd_inp" in the server instance.

I'm thinking of something like changing the behavior of the API with an option called `behavior` that has the choices `instruction`, `chat` and `generation`. There could also be an endpoint for performing embeddings and tokenization.

Comment (SlyEcho) [reply examples/server/server.cpp:699]: If the client is sending a string that starts with the same text as last time, there is no need to evaluate it again, tokenize the new string, find how many tokens are the same at the start and set that as `n_past`. Anyway, at this point with a GPU we can evaluate even 512 tokens in less than 10 seconds (depending on hardware).

For the streaming, maybe the easiest solution is to have no streaming, just return the generated text. If the client wants to show individual tokens appearing to the user, it could just limit `n_predict` to 1 and call the API in a loop.

Comment (FSSRepo) [reply examples/server/server.cpp:699]: I will try it. Good idea. I have to compare the new string tokens with `embd_inp`?

File: examples/CMakeLists.txt [hunk 4]
```diff
@@ -36,5 +36,6 @@ else()
     add_subdirectory(embedding)
     add_subdirectory(save-load-state)
     add_subdirectory(benchmark)
+    add_subdirectory(server)
```
Comment (ggerganov) [inline examples/CMakeLists.txt:39]: This has to be behind a CMake option `LLAMA_BUILD_SERVER`:

```cmake
    # in root CMakeLists.txt
    option(LLAMA_BUILD_SERVER "llama: build server example" OFF)

    # in examples/CMakeLists.txt
    if (LLAMA_BUILD_SERVER)
        add_subdirectory(server)
    endif()
```

**Review by ggerganov — CHANGES_REQUESTED** [review]

File: examples/CMakeLists.txt (hunk 4 above)
Comment (FSSRepo) [reply examples/CMakeLists.txt:39]: Ok

Comment (FSSRepo) [reply examples/CMakeLists.txt]: Done

Comment (FSSRepo) [reply examples/CMakeLists.txt]: I was thinking of including a small reference to this example in the README.md of the project, at least as a feature.

**Comment by howard0su** [comment]
If you can simulate OpenAI REST API, this example will become much more useful (even with some limitations). 
https://platform.openai.com/docs/api-reference/completions
https://platform.openai.com/docs/api-reference/embeddings

**Review by ggerganov — APPROVED** [review]
